### PR TITLE
Spreadsheet: Enable undo/redo of bucket color fill

### PR DESCRIPTION
**Pull request by huttongrabiel**
Undo and redo filling in color in Spreadsheet.

https://user-images.githubusercontent.com/60057590/229401577-02a31d77-05a4-42b0-9cc9-681c1033f6ad.mp4

### CODE REVIEW

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 1]
```diff
@@ -484,12 +484,17 @@ void SpreadsheetWidget::change_cell_static_color_format(Spreadsheet::FormatType
...
     auto dialog = GUI::ColorPicker::construct(Color::White, window(), "Select Color");
     if (dialog->exec() == GUI::Dialog::ExecResult::OK) {
+        Vector<CellChange> cell_changes;
         for (auto& position : current_worksheet_if_available()->selected_cells()) {
+            auto previous_type_metadata = current_worksheet_if_available()->at(position)->type_metadata();
             if (format_type == Spreadsheet::FormatType::Background)
                 current_worksheet_if_available()->at(position)->type_metadata().static_format.background_color = dialog->color();
             else
                 current_worksheet_if_available()->at(position)->type_metadata().static_format.foreground_color = dialog->color();
+            cell_changes.append(CellChange(*current_worksheet_if_available()->at(position), previous_type_metadata));
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:494]: Let's not lookup the cell three times per change.

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 2]
```diff
@@ -484,12 +484,17 @@ void SpreadsheetWidget::change_cell_static_color_format(Spreadsheet::FormatType
...
         }
+        undo_stack().push(make<CellsUndoMetadataCommand>(cell_changes));
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:496]: `move()` the `cell_changes` object here.

File: Userland/Applications/Spreadsheet/Spreadsheet.h [hunk 3]
```diff
@@ -25,15 +25,20 @@ namespace Spreadsheet {
...
     auto& new_data() { return m_new_data; }
+    auto& previous_type_metadata() { return m_previous_type_metadata; }
+    auto& new_type_metadata() { return m_new_type_metadata; }
 
 private:
     Cell& m_cell;
     DeprecatedString m_previous_data;
     DeprecatedString m_new_data;
+    CellTypeMetadata m_previous_type_metadata;
+    CellTypeMetadata m_new_type_metadata;
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/Spreadsheet.h:41]: Some of these fields are uninhabited with your change, I'd rather have only the available fields active (i.e. in a variant, or a union given that the discriminator is not actually within this class).

Comment (huttongrabiel) [reply Userland/Applications/Spreadsheet/Spreadsheet.h:41]: I'm not sure how to do this.

Would it be something like:
```cpp
struct Data {
	DeprecatedString previous_data;
	DeprecatedString new_data;
} Data;

struct TypeMetadata {
	CellTypeMetadata previous_type_metadata;
	CellTypeMetadata new_type_metadata;
} TypeMetadata;

Variant<Data, TypeMetadata> m_active_data;
```

Comment (alimpfard) [reply Userland/Applications/Spreadsheet/Spreadsheet.h:41]: I guess that could work, yeah.
this one isn't that important, you can ignore the comment if you want.

Comment (huttongrabiel) [reply Userland/Applications/Spreadsheet/Spreadsheet.h:41]: This one is making my head spin so I'm going to ignore for now and I'll open a new PR if I figure it out.

**Review by AtkinsSJ — APPROVED** [review]
Looks good! Not sure how this slipped through the cracks.

**Pull request merged**